kolena/_utils/instrumentation.py

```diff
@@ -87,3 +88,42 @@ def report_crash(id: int, endpoint_path: str):
     request = CoreAPI.MarkCrashedRequest(test_run_id=id)
     # note no krequests.raise_for_status -- already in crashed state
     krequests.post(endpoint_path=endpoint_path, data=json.dumps(dataclasses.asdict(request)))
+
+
+def set_profile():
+    krequests.put(endpoint_path=EventAPI.Path.PROFILE)
+
+
+def record_event(request: EventAPI.RecordEventRequest):
+    krequests.post(endpoint_path=EventAPI.Path.EVENT, json=dataclasses.asdict(request))
+
+
+def with_event(event_name: str):
+    """function decorator to track start and end of an event"""
+
+    def event_decorator(func: Callable) -> Callable:
+        @functools.wraps(func)
+        def wrapper(*args: Any, **kwargs: Any) -> Any:
+            if not _client_state.telemetry:
+                return func(*args, **kwargs)
+            # track duration of the call, and if failed record the exception class name
+            start_time = datetime.datetime.now()
+            event_metadata = {}
+            try:
+                response = func(*args, **kwargs)
+                return response
+            except Exception as e:
+                event_metadata["response_error"] = e.__class__.__name__
+                raise e
+            finally:
+                event_metadata["duration"] = round((datetime.datetime.now() - start_time).total_seconds(), 3)
+                record_event(
+                    EventAPI.RecordEventRequest(
+                        event_name=event_name,
+                        additional_metadata=event_metadata,
+                    ),
+                )
```

Comment:
Are there any cases where this can raise an exception and crash the main control flow?

Reply:
I do not expect this to happen given that the [server side swallows exceptions](https://github.com/kolenaIO/shipyard/blob/trunk/sdk-server/server/services/v1/tracking/service.py#L50,L73). At the same time I agree that there might be unexpected situations and it would be a good idea to swallow them on the client side as well.

Reply:
A good example would be changing the endpoint from `/tracking/event` to `/event` which will lead to errors if the server and client are not in sync briefly